### PR TITLE
Fix(optimizer): replace star with outer column list

### PR DESCRIPTION
**Pull request by georgesittas**

### CODE REVIEW

File: sqlglot/optimizer/scope.py [hunk 1]
```diff
@@ -79,6 +79,10 @@ def __init__(
         self.cte_scopes = []
         self.union_scopes = []
         self.udtf_scopes = []
+
+        if scope_type == ScopeType.UNION and parent and not outer_column_list:
```
Comment (barakalon) [inline sqlglot/optimizer/scope.py:83]: Any way to do this during scope traversal? This conditional, plus crawling back up the tree, feels hacky

Comment (georgesittas) [reply sqlglot/optimizer/scope.py:83]: I see what you mean, let me see if I can refactor it

Comment (georgesittas) [reply sqlglot/optimizer/scope.py]: Check out https://github.com/tobymao/sqlglot/pull/2776/commits/5d5b08bc95f73291eeb7ab4966bf6f27d8226d3b

**Review by barakalon — APPROVED** [review]
Nice 👍

**Pull request merged**